From 522c73bd41dce935524550ba0094cd33cc5f8e28 Mon Sep 17 00:00:00 2001
From: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@redhat.com>
Date: Thu, 28 May 2015 04:31:53 -0400
Subject: [PATCH] network: fix instance cache refresh for empty list

The bug introduce a race condition makes network_info as an
empty value. When trying to refresh cache we should use data
returned per Neutron.

(cherry picked from commit 4bb1e081bb87982f74c620b5b1f2376f8eb86640)

Co-Authored-By: Lee Yarwood <lyarwood@redhat.com>

Upstream-Liberty: https://review.openstack.org/#/c/189350/
Resolves: rhbz#1227352
Closes-Bug: #1407664
Change-Id: I3b9f60a2c4aea904f39d6914c8f35c4d682a6b3a
---
 nova/network/neutronv2/api.py             |  5 +++
 nova/tests/unit/network/test_neutronv2.py | 52 +++++++++++++++++++++++++++++++
 2 files changed, 57 insertions(+)

diff --git a/nova/network/neutronv2/api.py b/nova/network/neutronv2/api.py
index 1f96360..2f84715 100644
--- a/nova/network/neutronv2/api.py
+++ b/nova/network/neutronv2/api.py
@@ -1534,6 +1534,11 @@ class API(base_api.NetworkAPI):
             current_neutron_port_map[current_neutron_port['id']] = (
                 current_neutron_port)
 
+        # In that case we should repopulate ports from the state of
+        # Neutron.
+        if not port_ids:
+            port_ids = current_neutron_port_map.keys()
+
         for port_id in port_ids:
             current_neutron_port = current_neutron_port_map.get(port_id)
             if current_neutron_port:
diff --git a/nova/tests/unit/network/test_neutronv2.py b/nova/tests/unit/network/test_neutronv2.py
index 1f791cc..008b6c8 100644
--- a/nova/tests/unit/network/test_neutronv2.py
+++ b/nova/tests/unit/network/test_neutronv2.py
@@ -2665,6 +2665,58 @@ class TestNeutronv2(TestNeutronv2Base):
         self.assertFalse(nw_infos[4]['preserve_on_delete'])
         self.assertTrue(nw_infos[5]['preserve_on_delete'])
 
+    @mock.patch('nova.network.neutronv2.api.API._nw_info_get_subnets')
+    @mock.patch('nova.network.neutronv2.api.API._nw_info_get_ips')
+    @mock.patch('nova.network.neutronv2.api.API._nw_info_build_network')
+    @mock.patch('nova.network.neutronv2.api.API._get_preexisting_port_ids')
+    @mock.patch('nova.network.neutronv2.api.API._gather_port_ids_and_networks')
+    def test_build_network_info_model_empty(
+            self, mock_gather_port_ids_and_networks,
+            mock_get_preexisting_port_ids,
+            mock_nw_info_build_network,
+            mock_nw_info_get_ips,
+            mock_nw_info_get_subnets):
+        api = neutronapi.API()
+
+        fake_inst = objects.Instance()
+        fake_inst.project_id = 'fake'
+        fake_inst.uuid = 'uuid'
+        fake_inst.info_cache = objects.InstanceInfoCache()
+        fake_inst.info_cache.network_info = model.NetworkInfo()
+        fake_ports = [
+            # admin_state_up=True and status='ACTIVE' thus vif.active=True
+            {'id': 'port1',
+             'network_id': 'net-id',
+             'admin_state_up': True,
+             'status': 'ACTIVE',
+             'fixed_ips': [{'ip_address': '1.1.1.1'}],
+             'mac_address': 'de:ad:be:ef:00:01',
+             'binding:vif_type': model.VIF_TYPE_BRIDGE,
+             'binding:vnic_type': model.VNIC_TYPE_NORMAL,
+             'binding:vif_details': {},
+             },
+            ]
+        fake_subnets = [model.Subnet(cidr='1.0.0.0/8')]
+
+        neutronapi.get_client(mox.IgnoreArg(), admin=True).MultipleTimes(
+        ).AndReturn(self.moxed_client)
+        self.moxed_client.list_ports(
+            tenant_id='fake', device_id='uuid').AndReturn(
+                {'ports': fake_ports})
+
+        mock_gather_port_ids_and_networks.return_value = (None, None)
+        mock_get_preexisting_port_ids.return_value = []
+        mock_nw_info_build_network.return_value = (None, None)
+        mock_nw_info_get_ips.return_value = []
+        mock_nw_info_get_subnets.return_value = fake_subnets
+
+        self.mox.ReplayAll()
+        neutronapi.get_client('fake')
+
+        nw_infos = api._build_network_info_model(
+            self.context, fake_inst)
+        self.assertEqual(len(nw_infos), 1)
+
     def test_get_subnets_from_port(self):
         api = neutronapi.API()
 
